### PR TITLE
Add fog demo and transition fixes

### PR DESCRIPTION
**Pull request by karimnaaji**
- Skip unnecessary uniform code path when terrain is rendering to texture:
On a typical frame with terrain this was called ~200 times, where only 20 were necessary (~10%)
- Fix invalid transition evaluation:
When `delay` was undefined but not `now`, properties.begin and properties.end were assigned incorrect values, leading to an immediate transition interpolation. This was noticed when testing transitions on fog.
- Fix style.hasTransition() when fog is transitioning values
- Add a fog demo for release:
Split into two main animations, a fly to exercising complex expression zoom transitions, and a rotate exercising transitions.

https://user-images.githubusercontent.com/7061573/117220579-58c9a900-adbc-11eb-8da7-9619523e8913.mov

https://user-images.githubusercontent.com/7061573/117220376-f7093f00-adbb-11eb-8741-75e39529854a.mov

### CODE REVIEW

File: src/render/painter.js [hunk 1]
```diff
@@ -837,6 +837,13 @@ class Painter {
 
     prepareDrawProgram(context: Context, program: Program<*>, tileID: ?UnwrappedTileID) {
         const fog = this.style.fog;
+
+        // Fog isn not enabled when rendering to texture so we
```
Comment (asheemmamoowala) [inline src/render/painter.js:841]:
```suggestion
        // Fog is not enabled when rendering to texture so we
```

File: debug/fog-demo.html [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+            'sky-atmosphere-color': 'rgba(255, 255, 255, 0.2)',
+            'sky-opacity': 0,
+            'sky-opacity-transition': {'duration': 500}
+        }
+    });
+
+    map.setFog({
+        'range': ['interpolate', ['exponential', 0.1],
+            ['zoom'],
+            10, ['literal', [1000, 1200]],
```
Comment (rreusser) [inline debug/fog-demo.html:70]:
```suggestion
            10, ['literal', [10, 12]],
```

File: debug/fog-demo.html [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            12, ['literal', [-150, 1000]]],
```
Comment (rreusser) [inline debug/fog-demo.html:71]:
```suggestion
            12, ['literal', [-1.5, 10]]],
```

**Review by rreusser — APPROVED** [review]
Demo looks great! Do I understand the goal of debug pages correctly that the goal is a visual check to catch obvious visual glitches that would otherwise escape render tests or unit tests?

**Comment by karimnaaji** [comment]
>Do I understand the goal of debug pages correctly that the goal is a visual check to catch obvious visual glitches that would otherwise escape render tests or unit tests

Yes that, as it's always useful to check transitions/animations and render tests aren't well suited for such animations. But this is also for the https://docs.mapbox.com/mapbox-gl-js/example/ page to exemplify how to use fog.

**Pull request merged**